networking/src/Ntp/Packet.hs

```diff
@@ -4,92 +4,142 @@
 module Ntp.Packet
     ( NtpPacket (..)
     , ntpPacketSize
-    , mkCliNtpPacket
+    , mkNtpPacket
     , NtpOffset (..)
-    , evalClockOffset
+    , clockOffsetPure
+    , clockOffset
+    , realMcsToNtp
+    , ntpToRealMcs
     ) where
 
 
 import           Control.Lens (each, (^..))
 import           Control.Monad (replicateM_)
-import           Control.Monad.Trans (MonadIO (..))
 import           Data.Binary (Binary (..))
-import           Data.Binary.Get (getInt8, getWord32be, getWord8)
-import           Data.Binary.Put (putWord32be, putWord8)
+import           Data.Binary.Get (getInt8, getWord32be, getWord8, skip)
+import           Data.Binary.Put (putInt8, putWord32be, putWord8)
+import           Data.Int (Int8)
+import           Data.Time.Clock.POSIX (getPOSIXTime)
 import           Data.Time.Units (Microsecond, TimeUnit, fromMicroseconds,
                      toMicroseconds)
 import           Data.Word (Word32, Word8)
 
-import           Ntp.Util (getCurrentTime)
-
 data NtpPacket = NtpPacket
-    { ntpParams       :: Word8        -- some magic parameters
-    , ntpPoll         :: Int          -- poll delay between requests
-    , ntpOriginTime   :: Microsecond  -- when server sent reply
-    , ntpReceivedTime :: Microsecond  -- when server got request
-    , ntpTransmitTime :: Microsecond  -- when client sent request
-    } deriving (Show)
-
--- | NTP timestamp is not the same as time past from start of UNIX epoch — they
--- differ by 70 years.
-ntpTimestampDelta :: Integer
-ntpTimestampDelta = 2208988800
-
-ntpPacketSize :: Int
-ntpPacketSize = 48
-
-exp2'32 :: Integer
-exp2'32 = 4294967296
-
-ntpToRealMcs :: Word32 -> Word32 -> Microsecond
-ntpToRealMcs integerSec fracSec = fromMicroseconds $
-       (fromIntegral integerSec - ntpTimestampDelta) * 1000000
-      + ((fromIntegral fracSec * 1000000) `div` exp2'32)
-
-realMcsToNtp :: Microsecond -> (Word32, Word32)
-realMcsToNtp (toMicroseconds -> mcs) =
-    let integerSec = (mcs `div` 1000000) + ntpTimestampDelta
-        fracSec    = ((mcs * exp2'32) `div` 1000000)
-    in  (fromIntegral integerSec, fromIntegral fracSec)
+    { ntpParams       :: Word8        -- ^ some magic parameters
+    , ntpPoll         :: Int8         -- ^ poll delay between requests
+    , ntpOriginTime   :: Microsecond  -- ^ when server sent reply
+    , ntpReceivedTime :: Microsecond  -- ^ when server got request
+    , ntpTransmitTime :: Microsecond  -- ^ when client sent request
+    } deriving (Show, Eq)
 
 instance Binary NtpPacket where
     put NtpPacket{..} = do
         putWord8 ntpParams
+        putWord8 0
+        putInt8 ntpPoll
+        putWord8 0
```

Comment:
Not really, I added bits just so we have a test that `put` and `get` are inverse of each other.

Reply:
They weren't inverse prior to this change?
So I'm guessing the `replicateM_ 4 $ putWord8 0` was there beecause `ntpPoll` was assumed to be `0`. Here there's a `putInt8` rather than `putWord8`. Same encoding?
Then there's the `replicateM_ 9 $ PutWord32be 0`, which is replaced by `replicateM_ 5 $ putWord32be 0` but where are the remaining 4? `ntpOriginTime ^.. each` and `ntpReceivedTime ^.. each` give 2 `Word32`s each? Lens combinators are annoying hard to search for.